src/celeritas/optical/detail/OpticalGenAlgorithms.cu

```diff
@@ -82,6 +83,37 @@ count_num_photons(GeneratorDistributionRef<MemSpace::device> const& buffer,
     return count;
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the inclusive prefix sum of the number of optical primaries.
+ *
+ * \return Total accumulated value
+ */
+size_type inclusive_scan_primaries(
+    GeneratorDistributionRef<MemSpace::device> const& buffer,
+    Collection<size_type, Ownership::reference, MemSpace::device> const& offsets,
+    size_type size,
+    StreamId stream)
+{
+    CELER_EXPECT(!buffer.empty());
+    CELER_EXPECT(size > 0 && size <= buffer.size());
+    CELER_EXPECT(offsets.size() == buffer.size());
+
+    ScopedProfiling profile_this{"inclusive-scan-primaries"};
+    auto data = thrust::device_pointer_cast(buffer.data().get());
+    auto result = thrust::device_pointer_cast(offsets.data().get());
+    auto stop = thrust::transform_inclusive_scan(thrust_execute_on(stream),
+                                                 data,
+                                                 data + size,
+                                                 result,
+                                                 GetNumPhotons{},
+                                                 thrust::plus<size_type>());
+    CELER_DEVICE_CHECK_ERROR();
+
+    // Copy the last element (accumulated total) back to host
+    return *(stop - 1);
```

Comment:
Oh that is a very good question. I'm not positive, but I suspect it is synchronous...

We do this also in our core track init algorithms. A little while back I had refactored those to get rid of the `thrust::remove_if`, which @esseivaju had shown in his profiling requires synchronizing with the device. I didn't end up seeing much improvement, but this might explain why. I'm going to revisit that but using our async `Copier` to get the result of the scan instead...

Reply:
Cool! This is pretty subtle... and if this actually does improve performance/remove the synchronize, then maybe a little `Copier`-like class that simply returns a simple element (via an async copy) could be useful.